Allow to attach user-mode networking config to user's custom user-data file

lstocchi opened the pull request:
This PR contains the logic to merge a custom user data with the one we generate for user-mode networking.
It was part of https://github.com/cfergeau/podman/pull/31 but it was split to a different PR for easier review.

This commit was born after this discussion https://github.com/cfergeau/podman/pull/31#discussion_r2419212993

It should work this way

- custom user-data + system networking = we use user's user-data without any change and do not add any extra file to the iso
- NO custom user-data + system networking = we generate a default user-data having the core user/ssh set (no extra file is added to the iso)
- custom user-data + user-mode networking = we merge user's user-data with one created by us that contains what's needed to set/run gvforwarder + add gvforwarder binary in iso
- NO custom user-data + user-mode networking = we generate a default user-data having the core user/ssh set and gvforwarder stuff + gvforwarder binary in iso

lstocchi committed:
generate user-data config file based on settings on hyperv

if users provide their custom user-data we use theirc file. However,
user-mode networking requires some customization on the VM and we cannot
expect the user to provide such values. So, if users enable user-mode
networking and they provide a custom cloud-init, we update it to apply
the settings needed for user-mode networking to work fine.

if users provide a user-data file and they rely on system networking, we
keep their user-data as is.

If users do not provide any cloud init configuration, it creates a
default one.

Signed-off-by: lstocchi <lstocchi@redhat.com>

cfergeau commented:
Looks good to me, though I would prefer if this was split in 2 commits as in https://github.com/cfergeau/podman/pull/39 ([cloud-init: Refactor existing code](https://github.com/cfergeau/podman/pull/39/commits/effc29f1c244fc397f758e18eccccc414b46af16) + [cloud-init: generate user-data config file based on settings on hyperv](https://github.com/cfergeau/podman/pull/39/commits/bf48f3bdfe8a0331a170c209b4b782647fa7a153)).

I also have concerns about this kind of automatic modifications to a user-provided file, imo it’s more likely to break than not, and we won’t provide a lot of details regarding what macadam did. The code exists, so let’s try it, I’m happy to be proven wrong :)

lstocchi committed:
remove newline from beginning HyperVVsockNMConnection

it removes the initial newline char from the string
HyperVVsockNMConnection.
The newline creates no issues for ignition but break the configuration
when using cloud-init. This fixes the problem.

lstocchi committed:
define cloud-init merging strategy

when merging two cloud-init config files like the case where one user-data is passed by the user and one is
created by us to support user-mode networking, we need to make sure we
define a merging strategy, otherwise cloud-init will just use the last
clause by overwriting the old one.

For example if there are two files like

runcmd:
  - bash1
  - bash2

runcmd:
  - bash3
  - bash4

if we do not define a merging strategy, cloud-init will only run bash3
and bash4.

This could be an issue if user passes a user-data file containing same
clauses we use in our user-mode networking config file (like runcmd).

This path defines a strategy to merge lists of value and do not do
anything with string

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi committed:
use actual path where gvforwarder is copied by cloud-init

Signed-off-by: lstocchi <lstocchi@redhat.com>

cfergeau commented:
This was included in https://github.com/cfergeau/podman/pull/39 which is now merged.
